From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MeFisto94 <MeFisto94@users.noreply.github.com>
Date: Fri, 20 Aug 2021 00:09:35 +0200
Subject: [PATCH] Added an overload of Raft_Network#TryToJoinGame that takes an
 IP-Port Pair and added calls to the command line parsing for connect and
 server hosting


diff --git a/Raft_Network.cs b/Raft_Network.cs
index 2c923519f9be91999deea3f0e1d58a299d41009b..6131fa44176bf1894d22df77e6f2dc3d6824d878 100644
--- a/Raft_Network.cs
+++ b/Raft_Network.cs
@@ -937,7 +937,13 @@ public class Raft_Network : MonoBehaviour, SceneEventInterface
 		}
 	}
 
-	public void TryToJoinGame(CSteamID hostID, string password)
+	// Sharkkit Dedicated Start - Move method body down into a method also accepting ip:port addresses.
+	public void TryToJoinGame(CSteamID hostId, string password)
+	{
+		TryToJoinGame(hostId, password, null, 0);
+	}
+
+	public void TryToJoinGame(CSteamID hostId, string password, string ipAddress, ushort port)
 	{
 		if (IsConnectedToHost)
 		{
@@ -945,15 +951,27 @@ public class Raft_Network : MonoBehaviour, SceneEventInterface
 		}
 		else if (!awaitingConnectionResponseUser.IsValid())
 		{
-			this.hostID = hostID;
-			awaitingConnectionResponseUser = hostID;
+			this.hostID = hostId;
+			awaitingConnectionResponseUser = hostId;
 			isHost = false;
 			gameManagerStartCalled = false;
 			worldLoadedCoroutine = null;
-			Helper.LogBuild("Sending request to join " + SteamFriends.GetFriendPersonaName(hostID) + "'s game");
-			SendP2P(hostID, new Message_InitiateConnection(Messages.InitiateSession, Settings.AppBuildID, password), EP2PSend.k_EP2PSendReliable, NetworkChannel.Channel_Session);
+
+			if (ipAddress != null)
+			{
+				// Sharkkit Dedicated - The Network Wrapper needs to prepare a connection
+				NetworkWrapper.ConnectToIP(hostId, ipAddress, port);
+				Helper.LogBuild($"Sending request to join the dedicated server at {ipAddress}:{port}");
+			}
+			else
+			{
+				Helper.LogBuild("Sending request to join " + SteamFriends.GetFriendPersonaName(hostId) + "'s game");
+			}
+
+			SendP2P(hostId, new Message_InitiateConnection(Messages.InitiateSession, Settings.AppBuildID, password), EP2PSend.k_EP2PSendReliable, NetworkChannel.Channel_Session);
 		}
 	}
+	// Sharkkit Dedicated End - Move method body down into a method also accepting ip:port addresses.
 
 	public void LeaveGame(DisconnectReason reason, SceneName sceneName, bool saveGame, bool force = false)
 	{
@@ -991,8 +1009,7 @@ public class Raft_Network : MonoBehaviour, SceneEventInterface
 	{
 		if (awaitingConnectionResponseUser.IsValid())
 		{
-			// Sharkkit Dedicated - Redirect call to Network Wrapper
-			NetworkWrapper.CloseP2PSessionWithUser(HostID);
+			NetworkWrapper.CloseP2PSessionWithUser(HostID); // Sharkkit Dedicated - Redirect call to Network Wrapper
 			awaitingConnectionResponseUser.Clear();
 			hostID.Clear();
 		}
@@ -1321,13 +1338,36 @@ public class Raft_Network : MonoBehaviour, SceneEventInterface
 		{
 			return;
 		}
-		foreach (string text2 in commandLineArgs)
+
+		// Sharkkit Dedicated Start - Parse the dedicated commandLine and host a server
+		if (CommandLine.ParseDedicated(Array.AsReadOnly(commandLineArgs), out var listenArguments))
+		{
+			// TODO: Delay call to ListenOn until after a world has been loaded, to prevent crashes when connecting too early
+			NetworkWrapper.ListenOn(listenArguments);
+		}
+		// Sharkkit Dedicated End - Parse the dedicated commandLine and host a server
+
+		foreach (var arg in commandLineArgs)
 		{
-			text += text2;
-			if (text2.Contains(CONNECT))
+			text += arg;
+			if (arg.Contains(CONNECT))
 			{
-				HandleConnectArgument(text2);
+				HandleConnectArgument(arg);
+			}
+			// Sharkkit Dedicated Start - Parse ips from the commandline and auto-connect
+			else if (CommandLine.ParseIpConnect(arg, out var addr, out var port, out var steamId))
+			{
+				// The following is a copy of the original join code, but calling the correct TryToJoinGame Method
+				var startMenuScreen = FindObjectOfType<StartMenuScreen>();
+				// TODO: Password
+				TryToJoinGame(steamId, string.Empty, addr, port);
+				if (startMenuScreen != null && startMenuScreen.joinGameBox != null)
+				{
+					startMenuScreen.joinGameBox.connectingBox.gameObject.SetActive(true);
+					startMenuScreen.joinGameBox.connectingBox.StartConnectTimeoutTimer();
+				}
 			}
+			// Sharkkit Dedicated End - Parse ips from the commandline and auto-connect
 		}
 		string[] array = text.Replace(" ", "").ToLower().Split('-');
 		foreach (string text3 in array)
